### ISSUE TITLE
s6-notifyoncheck - invalid argument

### ISSUE THREAD

**Issue by mwmason** [just-containers/s6-overlay#444]
Hi,

I'm trying to set up dependencies using s6-notifyoncheck.  I've defined a check in servicedir/data and in my run for the service I have:

```
#!/command/execlineb -P
s6-envuidgid redis 
s6-notifyoncheck -d -3 3 redis-server --port 6379 --bind 127.0.0.1 
```

I keep getting `s6-notifyoncheck: fatal: unable to sanity-check current service directory: Invalid argument` 

I have no idea what is wrong.  I've tried various arguments, no arguments, renaming the servicedir, reducing my data/check to just exit, including a notification-fd, using the -3 3 as above; using -d, not using -d.

Any ideas?

**Comment by skarnet** [comment]
This can only happen if [something's wrong](https://github.com/skarnet/s6/blob/master/src/supervision/s6-notifyoncheck.c#L136) with the `supervise/lock` file in your service directory: this is definitely not a normal condition.

Can you strace the s6-notifyoncheck call (add `strace` before s6-notifyoncheck in your run script,
and install the strace package in your image) and post the result? Please also `ls -l supervise/` as root in your service directory. Those two pieces of information will tell exactly what's going on.

**Comment by mwmason** [comment]
Ok, results below, ls -l supervise is first:

```
2022-05-15 02:15:52.308 EDT total 8
2022-05-15 02:15:52.308 EDT prw------- 1 root root 0 May 15 06:15 control
2022-05-15 02:15:52.308 EDT -rw-r--r-- 1 root root 182 May 15 06:15 death_tally
2022-05-15 02:15:52.308 EDT -rw-r--r-- 1 root root 0 May 15 06:15 lock
2022-05-15 02:15:52.308 EDT -rw-r--r-- 1 root root 35 May 15 06:15 status
2022-05-15 02:15:52.320 EDT Container Sandbox: Unsupported syscall process_vm_readv(0xb4,0x3e81b395ea28,0x1,0x3e81b395ea38,0x1,0x0). It is very likely that you can safely ignore this message and that this is not the cause of any error you might be troubleshooting. Please, refer to https://gvisor.dev/docs/user_guide/compatibility/linux/amd64/#process_vm_readv for more information.
2022-05-15 02:15:52.321 EDT execve("/command/s6-notifyoncheck", ["s6-notifyoncheck", "-d", "-3", "3", "s6-setuidgid", "redis", "redis-server", "--port", "6379", "--bind", "127.0.0.1"], 0x3e81b395fdb0 /* 3 vars */) = 0
2022-05-15 02:15:52.321 EDT arch_prctl(ARCH_SET_FS, 0x60c9f8) = 0
2022-05-15 02:15:52.321 EDT set_tid_address(0x60cb90) = 180
2022-05-15 02:15:52.321 EDT open("./supervise/lock", O_RDONLY|O_NONBLOCK|O_LARGEFILE) = 3
2022-05-15 02:15:52.321 EDT fcntl(3, F_GETLK, 0x3e5cbe107a20) = -1 EINVAL (Invalid argument)
2022-05-15 02:15:52.321 EDT close(3) = 0
2022-05-15 02:15:52.322 EDT writev(2, [{iov_base="s6-notifyoncheck: fatal: unable "..., iov_len=92}, {iov_base=NULL, iov_len=0}], 2s6-notifyoncheck: fatal: unable to sanity-check current service directory: Invalid argument
2022-05-15 02:15:52.322 EDT ) = 92
2022-05-15 02:15:52.322 EDT exit_group(111) = ?
2022-05-15 02:15:52.322 EDT +++ exited with 111 +++
```

**Comment by skarnet** [comment]
Well, it's `fcntl` failing for some reason.

It looks like you're using gVisor, not Linux, and the gVisor compatibility page for fcntl says [Partial support](https://gvisor.dev/docs/user_guide/compatibility/linux/amd64/#fcntl). The most probable explanation is that gVisor does not support the F_GETLK operation for fcntl, which makes it non-compliant with POSIX.

My advice is to check with gVisor people. If they confirm that their implementation of fcntl does not support F_GETLK, then I'm afraid s6-notifyoncheck, as well as other parts of s6 that use file locking, is simply unusable on gVisor.

I'm sorry, you're not the first person to fall for yet another fake-ass emulation.

**Comment by mwmason** [comment]
Well, this is a relief - I really thought I was doing something incorrectly.  

So I guess I can go back to my original approach of using s6-sleep,  or I could implement a poll using the check I already wrote.  

Thank you again!

**Comment by skarnet** [comment]
I have been [told](https://github.com/google/gvisor/issues/5113) that gVisor has implemented support for F_GETLK, but only with its [vfs2](https://github.com/google/gvisor/issues/5113#issuecomment-765880268) option. Maybe you can use s6-notifyoncheck if you make sure to run gVisor with -vfs2?

**Comment by mwmason** [comment]
Well, there's no exposed option to choose -vfs2, but apparently selecting [Preview - Second Generation](https://stackoverflow.com/questions/69846927/how-to-run-untrusted-code-using-gvisor-on-google-cloud-run) accomplishes the same.

While I get a different result, it's still an error - Bad file descriptor.  I tried '101' just to see, and the same result.

```
2022-05-15 23:20:55.554 EDT execve("/command/s6-notifyoncheck", ["s6-notifyoncheck", "-d", "-3", "3", "s6-setuidgid", "redis", "redis-server", "--port", "6379", "--bind", "127.0.0.1"], 0x7fff217473f0 /* 3 vars */) = 0
2022-05-15 23:20:55.554 EDT arch_prctl(ARCH_SET_FS, 0x60c9f8) = 0
2022-05-15 23:20:55.554 EDT set_tid_address(0x60cb90) = 80
2022-05-15 23:20:55.554 EDT open("./supervise/lock", O_RDONLY|O_NONBLOCK|O_LARGEFILE) = 3
2022-05-15 23:20:55.554 EDT fcntl(3, F_GETLK, {l_type=F_WRLCK, l_whence=SEEK_SET, l_start=0, l_len=0, l_pid=23}) = 0
2022-05-15 23:20:55.554 EDT close(3) = 0
2022-05-15 23:20:55.554 EDT fcntl(3, F_GETFD) = -1 EBADF (Bad file descriptor)
2022-05-15 23:20:55.554 EDT writev(2, [{iov_base="s6-notifyoncheck: fatal: notific"..., iov_len=82}, {iov_base=NULL, iov_len=0}], 2s6-notifyoncheck: fatal: notification-fd sanity check failed: Bad file descriptor
2022-05-15 23:20:55.554 EDT ) = 82
2022-05-15 23:20:55.555 EDT exit_group(111) = ?
2022-05-15 23:20:55.555 EDT +++ exited with 111 +++
```
Thanks for restoring my sanity - I can accept it's a limitation of my environment and my sleep workaround will have to suffice - just not as elegant.

**Comment by skarnet** [comment]
Glad you could make it work!
That's a totally different error you're getting here. :-) In order for `s6-notifyoncheck` to work, you need to have a `notification-fd` file in your service directory, containing the fd number you're using: here, 3, since you're using `-3 3`.

`echo 3 > notification-fd` in your service directory should solve your issue.

**Comment by mwmason** [comment]
Oops! Guess I thought notification-fd was not needed if I used the -3 option - thanks for highlighting that!

Good news is - it's all working with dependencies and checks defined (and no s6-sleeps)!  

I did have to up S6_CMD_WAIT_FOR_SERVICES_MAXTIME because 5000 ms wasn't enough time to allow the last service - with its dependencies - time to start.  

But after getting this all configured, it's working like a charm - thank you again for all of your commitment and support of this project.
